main.py

```diff
@@ -0,0 +1,277 @@
+            {
+                "losses": losses.item(),
+                "epoch": epoch,
+                "lr": optimizer.param_groups[0]["lr"],
+                **{key: value.item() for key, value in loss_dict.items()},
+            }
+        )
+
+
+def train_loop_per_worker(config):
+    model = models.detection.fasterrcnn_resnet50_fpn(num_classes=21)
+    model = train.torch.prepare_model(model)
+    parameters = [p for p in model.parameters() if p.requires_grad]
+    optimizer = torch.optim.SGD(
+        parameters,
+        lr=config["lr"],
+        momentum=config["momentum"],
+        weight_decay=config["weight_decay"],
+    )
+    lr_scheduler = torch.optim.lr_scheduler.MultiStepLR(
+        optimizer, milestones=config["lr_steps"], gamma=config["lr_gamma"]
+    )
+    start_epoch = 0
+
+    checkpoint = session.get_checkpoint()
+    if checkpoint is not None:
+        checkpoint_data = checkpoint.to_dict()
+        model.load_state_dict(checkpoint_data["model"])
+        optimizer.load_state_dict(checkpoint_data["optimizer"])
+        lr_scheduler.load_state_dict(checkpoint_data["lr_scheduler"])
+        start_epoch = checkpoint_data["epoch"]
+
+    for epoch in range(start_epoch, config["epochs"]):
+        train_one_epoch(
+            model=model,
+            optimizer=optimizer,
+            batch_size=config["batch_size"],
+            epoch=epoch,
+        )
+        lr_scheduler.step()
+        checkpoint = Checkpoint.from_dict(
+            {
+                "model": model.module.state_dict(),
+                "optimizer": optimizer.state_dict(),
+                "lr_scheduler": lr_scheduler.state_dict(),
+                "config": config,
+                "epoch": epoch,
+            }
+        )
+        session.report({}, checkpoint=checkpoint)
+
+
+trainer = TorchTrainer(
+    train_loop_per_worker=train_loop_per_worker,
+    train_loop_config={
+        "batch_size": 2,
+        "lr": 0.02,
+        # "epochs": 26,
+        "epochs": 1,
+        "momentum": 0.9,
+        "weight_decay": 1e-4,
+        "lr_steps": [16, 22],
+        "lr_gamma": 0.1,
+    },
+    scaling_config=ScalingConfig(num_workers=2, use_gpu=False),
+    # run_config=RunConfig(progress_reporter=CLIReporter)  TODO
+    datasets={"train": train_dataset.limit(100)},
+)
+results = trainer.fit()
+
+
+class CustomTorchPredictor(TorchPredictor):
+    def _predict_numpy(
+        self, data: np.ndarray, dtype: torch.dtype
+    ) -> Dict[str, np.ndarray]:
+        inputs = [torch.as_tensor(image) for image in data["image"]]
+        assert all(image.dim() == 3 for image in inputs)
+        outputs = self.call_model(inputs)
+
+        predictions = collections.defaultdict(list)
+        for output in outputs:
+            for key, value in output.items():
+                predictions[key].append(value.cpu().detach().numpy())
+
+        for key, value in predictions.items():
+            predictions[key] = _create_possibly_ragged_ndarray(value)
```

Comment:
is the _create_possibly_ragged_ndarray needed here? there should be automatic support for ragged tensors right?

Reply:
Yeah. Otherwise you get errors like this:
```
ValueError: ArrowVariableShapedTensorArray only supports heterogeneous-shaped tensor collections, not arbitrarily nested ragged tensors. Got arrays: [('dtype=object', 'shape=(100, 4)')]
```